src/index.js

```diff
@@ -34,18 +34,109 @@ function registerJavaScriptInjections(hljs) {
 
   js = js.rawDefinition(hljs);
 
+  setupHBSLiteral(js);
+  swapXMLForGlimmer(js);
+  setupTemplateTag(js);
+
+  hljs.registerLanguage('javascript', () => js);
+}
+
+function setupHBSLiteral(js) {
   let cssIndex = js.contains.findIndex((rule) => rule?.begin === 'css`');
   let css = js.contains[cssIndex];
 
+  const HBS_TEMPLATE = hljs.inherit(css, { begin: /hbs`/ });
+
+  HBS_TEMPLATE.starts.subLanguage = 'glimmer';
+
+  js.contains.splice(cssIndex, 0, HBS_TEMPLATE);
+}
+
+function swapXMLForGlimmer(js) {
   // The default JSX grammar is actually just XML, which... is also wrong :D
   js.contains
     .flatMap((contains) => contains?.contains || contains)
     .filter((rule) => rule.subLanguage === 'xml')
     .forEach((rule) => (rule.subLanguage = 'glimmer'));
+}
 
-  const HBS_TEMPLATE = hljs.inherit(css, { begin: /hbs`/ });
+/**
+ * A lot of this is stolen from XML
+ */
+function setupTemplateTag(js) {
+  const GLIMMER_TEMPLATE_TAG = {
+    begin: /<template>/,
+    end: /<\/template>/,
+    /**
+     * @param {RegExpMatchArray} match
+     * @param {CallbackResponse} response
+     */
+    isTrulyOpeningTag: (match, response) => {
+      const afterMatchIndex = match[0].length + match.index;
+      const nextChar = match.input[afterMatchIndex];
 
-  HBS_TEMPLATE.starts.subLanguage = 'glimmer';
-  js.contains.splice(cssIndex, 0, HBS_TEMPLATE);
-  hljs.registerLanguage('javascript', () => js);
+      if (
+        // HTML should not include another raw `<` inside a tag
+        // nested type?
+        // `<Array<Array<number>>`, etc.
+        nextChar === '<' ||
+        // the , gives away that this is not HTML
+        // `<T, A extends keyof T, V>`
+        nextChar === ','
+      ) {
+        response.ignoreMatch();
+
+        return;
+      }
+
+      // `<something>`
+      // Quite possibly a tag, lets look for a matching closing tag...
+      if (nextChar === '>') {
+        // if we cannot find a matching closing tag, then we
+        // will ignore it
+        if (!hasClosingTag(match, { after: afterMatchIndex })) {
+          response.ignoreMatch();
+        }
+      }
+
+      // `<blah />` (self-closing)
+      // handled by simpleSelfClosing rule
+
+      // `<From extends string>`
+      // technically this could be HTML, but it smells like a type
+      let m;
+      const afterMatch = match.input.substring(afterMatchIndex);
+
+      // NOTE: This is ugh, but added specifically for https://github.com/highlightjs/highlight.js/issues/3276
+      if ((m = afterMatch.match(/^\s+extends\s+/))) {
+        if (m.index === 0) {
+          response.ignoreMatch();
+
+          // eslint-disable-next-line no-useless-return
+          return;
+        }
+      }
+    },
+  };
+
+  js.contains.unshift({
```

Comment:
@joshgoebel this is how I'm setting up the new injection.

I guess, what _I really want_ is some type-safe/jsdoc/intellisense-aware way or utility to get intellisense for what APIs I have access to, and what the shape of the grammar is that I'm modifying. even if it's a no-op wrapper function that adds jsdoc / type-info, that'd be a big help.

Reply:
The HLJS API itself has defined TypeScript type info, but there is no real "shape" to grammars (beyond the basic keys which are both typed and documented), and there are many different ways to write them... any type of extensions is going to have to hook into *very* specific points - and even then would possibly still be subject to the grammar changing out from under it.

Again, see typescript:

```
  // TS gets a simpler shebang rule than JS
  swapMode(tsLanguage, "shebang", hljs.SHEBANG());
  // JS use strict rule purposely excludes `asm` which makes no sense
  swapMode(tsLanguage, "use_strict", USE_STRICT);
```

Even to do those two simple changes the JS grammar first had to "tag" those two modes with "shebang" and "use_strict" - making it easy for the TS grammar to "find" them in the haystack that is the whole JS grammar... coming entirely from the outside you'd have to do your best to do this every release - or perhaps use some sort of heuristics to find the areas you wanted to patch in, etc.